File: E-Commerce App Project (Tabbed)/Controllers/CheckOutViewController.m
//
//  CheckOutViewController.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 16/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "CheckOutViewController.h"

#define A4PaperSize CGRectMake(0,0,595.2,841.8);

@interface CheckOutViewController (){
    NSUserDefaults *defaults;
    NSDictionary *userData;
    NSString *PDFFilePath;
    UIAlertController *paymentAlert;
}

@end

@implementation CheckOutViewController

@synthesize checkOutTotalAmount,firstName,lastName,email,phone,country,city,State,postalCode,address,bkashOutlet,rocketOutlet,cashOnDeliveryOutlet;
@synthesize bkashChecked,rocketChecked,cashOnDeliveryChecked;

- (void)viewDidLoad {
    [super viewDidLoad];
    self.checkoutCart = [ShoppingCart sharedInstance];
    
    
    paymentAlert = [UIAlertController
                    alertControllerWithTitle:@"Payment Method Not Selected"
                    message:@"Please Choose your payment method From the Payment Method Section"
                    preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction* yesAction = [UIAlertAction
                                actionWithTitle:@"OK!"
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    
                                }];
    [paymentAlert addAction:yesAction];
    
    bkashChecked = NO;
    rocketChecked = NO;
    cashOnDeliveryChecked = NO;
    
    defaults = [NSUserDefaults standardUserDefaults];
    
    if (![defaults boolForKey:@"SeesionUserLoggedIN"]) {
        checkOutTotalAmount.text=@"";
        firstName.text=@"";
        lastName.text=@"";
        email.text=@"";
        phone.text=@"";
        country.text=@"";
        city.text=@"";
        State.text=@"";
        postalCode.text=@"";
        address.text=@"";
        bkashOutlet.hidden=YES;
        rocketOutlet.hidden=YES;
        cashOnDeliveryOutlet.hidden=YES;
    }else{
        bkashOutlet.hidden=NO;
        rocketOutlet.hidden=NO;
        cashOnDeliveryOutlet.hidden=NO;
        userData = [defaults dictionaryForKey:@"LoggedInUsersDetail"];
        
        checkOutTotalAmount.text = [NSString stringWithFormat:@"$%@",[defaults valueForKey:@"cartTotalAmount"]];
        firstName.text=[userData valueForKey:@"firstName"];
        lastName.text=[userData valueForKey:@"lastName"];
        email.text=[userData valueForKey:@"usersEmail"];
        phone.text=[userData valueForKey:@"phone"];
        country.text=[userData valueForKey:@"country"];
        city.text=[userData valueForKey:@"city"];
        State.text=[userData valueForKey:@"state"];
        postalCode.text=[userData valueForKey:@"postalCode"];
        address.text=[userData valueForKey:@"address"];
    }
    
    // Do any additional setup after loading the view.
    // NS userDefaults has all the values needed
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (IBAction)bkashAction:(id)sender {
    if (!bkashChecked) {
        [bkashOutlet setImage:[UIImage imageNamed:@"checkboxChecked-icon-40.png"] forState:UIControlStateNormal];
        [rocketOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [cashOnDeliveryOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        bkashChecked = YES;
        rocketChecked = NO;
        cashOnDeliveryChecked = NO;
        [defaults setObject:@"Bkash" forKey:@"paymentMethodUsed"];
        [defaults synchronize];
    }else if(bkashChecked){
        [bkashOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [rocketOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [cashOnDeliveryOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        bkashChecked = NO;
        rocketChecked = NO;
        cashOnDeliveryChecked = NO;
        [defaults setObject:@"" forKey:@"paymentMethodUsed"];
        [defaults synchronize];
    }
}

- (IBAction)rocketAction:(id)sender {
    if (!rocketChecked) {
        [rocketOutlet setImage:[UIImage imageNamed:@"checkboxChecked-icon-40.png"] forState:UIControlStateNormal];
        [bkashOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [cashOnDeliveryOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        rocketChecked = YES;
        bkashChecked = NO;
        cashOnDeliveryChecked = NO;
        [defaults setObject:@"Rocket" forKey:@"paymentMethodUsed"];
        [defaults synchronize];
    }else if(rocketChecked){
        [bkashOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [rocketOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [cashOnDeliveryOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        bkashChecked = NO;
        rocketChecked = NO;
        cashOnDeliveryChecked = NO;
        [defaults setObject:@"" forKey:@"paymentMethodUsed"];
        [defaults synchronize];
    }
}

- (IBAction)cashOnDeliveryAction:(id)sender {
    if (!cashOnDeliveryChecked) {
        [cashOnDeliveryOutlet setImage:[UIImage imageNamed:@"checkboxChecked-icon-40.png"] forState:UIControlStateNormal];
        [rocketOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [bkashOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        cashOnDeliveryChecked = YES;
        rocketChecked = NO;
        bkashChecked = NO;
        [defaults setObject:@"Cash On Delivery" forKey:@"paymentMethodUsed"];
        [defaults synchronize];
    }else if(cashOnDeliveryChecked){
        [bkashOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [rocketOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        [cashOnDeliveryOutlet setImage:[UIImage imageNamed:@"checkboxUnchecked-icon-40.png"] forState:UIControlStateNormal];
        bkashChecked = NO;
        rocketChecked = NO;
        cashOnDeliveryChecked = NO;
        [defaults setObject:@"" forKey:@"paymentMethodUsed"];
        [defaults synchronize];
    }
}
-(NSString*)getHTMLString{
   
    NSString *pdfLogoUrl = @"https://apiforios.appendtech.com/logo.png";
    
    NSString *filePath = [[NSBundle mainBundle] pathForResource:@"invoice" ofType:@"html"];
    NSString *singleItemFilePath = [[NSBundle mainBundle] pathForResource:@"single_item" ofType:@"html"];
    
    NSString *strHTML = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
    NSArray *itemsInCartArray = self.checkoutCart.itemsInCart;
    
    NSMutableArray *allItemsHTMLArray=[[NSMutableArray alloc]init];
    NSString *allItemsHTMLString;
    for (int i=0; i<itemsInCartArray.count; i++) {
        NSString *srtItemHTML = [NSString stringWithContentsOfFile:singleItemFilePath encoding:NSUTF8StringEncoding error:nil];
        srtItemHTML = [srtItemHTML stringByReplacingOccurrencesOfString:@"#productName" withString:[itemsInCartArray[i] itemName]];
        srtItemHTML = [srtItemHTML stringByReplacingOccurrencesOfString:@"#quantity" withString:[[NSNumber numberWithDouble:[itemsInCartArray[i] cartAddedQuantity]] stringValue]];
        srtItemHTML = [srtItemHTML stringByReplacingOccurrencesOfString:@"#price" withString:[[NSNumber numberWithDouble:[itemsInCartArray[i] price]] stringValue]];
        
        [allItemsHTMLArray addObject:srtItemHTML];
    }
    allItemsHTMLString = [allItemsHTMLArray componentsJoinedByString:@"\n"];
    
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#ITEMS#" withString:allItemsHTMLString];
    
    NSDate *date= [NSDate date];
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc]init];
    [dateFormatter setDateFormat:@"MMM d, yyyy"];
    
    NSString *dateString = [dateFormatter stringFromDate:date];
    
    NSString *ownerInfo = @"ICTcom<br>An E-commerce App Project<br>Developed BY: Rony Banik (Arko)<br>";
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#ownerInfo" withString:ownerInfo];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#appIcon" withString:pdfLogoUrl];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#invoiceDate" withString:dateString];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#cartTotal" withString:[defaults valueForKey:@"cartTotalAmount"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#firstName" withString:[userData valueForKey:@"firstName"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#lastName" withString:[userData valueForKey:@"lastName"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#email" withString:[userData valueForKey:@"usersEmail"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#phone" withString:[userData valueForKey:@"phone"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#country" withString:[userData valueForKey:@"country"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#city" withString:[userData valueForKey:@"city"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#state" withString:[userData valueForKey:@"state"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#postalCode" withString:[userData valueForKey:@"postalCode"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#address" withString:[userData valueForKey:@"address"]];
    strHTML = [strHTML stringByReplacingOccurrencesOfString:@"#paymentMethod" withString:[defaults valueForKey:@"paymentMethodUsed"]];
    
    return strHTML;
}

-(void)savePDF{
    CGRect pageFrame = A4PaperSize;
    
    UIPrintPageRenderer *printPageRenderer = [[UIPrintPageRenderer alloc]init];
    [printPageRenderer setValue:[NSValue valueWithCGRect:pageFrame] forKey:@"paperRect"];
    [printPageRenderer setValue:[NSValue valueWithCGRect:pageFrame] forKey:@"printableRect"];
    UIPrintFormatter *printFormatter = [[UIMarkupTextPrintFormatter alloc]initWithMarkupText:[self getHTMLString]];
    [printPageRenderer addPrintFormatter:printFormatter startingAtPageAtIndex:0];
    
    NSMutableData *pdfData = [[NSMutableData alloc]init];
    
    UIGraphicsBeginPDFContextToData(pdfData, CGRectZero, nil);
    UIGraphicsBeginPDFPage();
    [printPageRenderer drawPageAtIndex:0 inRect:UIGraphicsGetPDFContextBounds()];
    UIGraphicsEndPDFContext();
    PDFFilePath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, true)objectAtIndex:0];
    PDFFilePath = [PDFFilePath stringByAppendingString:@"/invoiceFromICTcom.pdf"];
    [pdfData writeToFile:PDFFilePath atomically:true];
    NSLog(@"Filepath: %@",PDFFilePath);
    [defaults setObject:PDFFilePath forKey:@"pdfFilePath"];
    [defaults synchronize];
}
- (IBAction)finishCheckOutButton:(id)sender {
    if ([[defaults valueForKey:@"paymentMethodUsed"] isEqualToString:@""]){
        [self presentViewController:paymentAlert animated:YES completion:nil];
    }else{
        [self savePDF];
        [self sendMailWithPDF];
        [defaults setObject:@"" forKey:@"cartTotalAmount"];
        [defaults setObject:@"" forKey:@"paymentMethodUsed"];
        [defaults synchronize];
        [self.checkoutCart clearCart];
        [self performSegueWithIdentifier:@"checkoutToThankyouSegue" sender:nil];
    }
}

-(void)sendMailWithPDF{
    if ([MFMailComposeViewController canSendMail])
    {
        MFMailComposeViewController *saleMail = [[MFMailComposeViewController alloc] init];
        saleMail.mailComposeDelegate = self;
        [saleMail setSubject:@"Invoice From ICTcom App"];
        [saleMail setMessageBody:@"Thanks so much for your purchase.<br>Please find the attached invoice.<br>Thanks From ICTcom" isHTML:YES];
        NSData *contentsOfFile = [[NSData alloc]initWithContentsOfFile:PDFFilePath];
        
        [saleMail addAttachmentData:contentsOfFile mimeType:@"application/pdf" fileName:@"Invoice"];
        
        [saleMail setToRecipients:[userData valueForKey:@"usersEmail"]];  //Set a test email
        
        [self presentViewController:saleMail animated:YES completion:NULL];
    }
    else
    {
        NSLog(@"Device can't send email! maybe Simulator");
    }
}
@end

File: E-Commerce App Project (Tabbed)/Controllers/CheckOutViewController.swift
//
//  CheckOutViewController.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift
//

import UIKit
import MessageUI

class CheckOutViewController: UIViewController, MFMailComposeViewControllerDelegate {
    
    @IBOutlet var checkOutTotalAmount: UILabel!
    @IBOutlet var firstName: UILabel!
    @IBOutlet var lastName: UILabel!
    @IBOutlet var email: UILabel!
    @IBOutlet var phone: UILabel!
    @IBOutlet var country: UILabel!
    @IBOutlet var city: UILabel!
    @IBOutlet var State: UILabel!
    @IBOutlet var postalCode: UILabel!
    @IBOutlet var address: UILabel!
    @IBOutlet var bkashOutlet: UIButton!
    @IBOutlet var rocketOutlet: UIButton!
    @IBOutlet var cashOnDeliveryOutlet: UIButton!
    
    var bkashChecked: Bool = false
    var rocketChecked: Bool = false
    var cashOnDeliveryChecked: Bool = false
    
    var checkoutCart: ShoppingCart!
    
    private var defaults: UserDefaults!
    private var userData: [String: Any]?
    private var PDFFilePath: String = ""
    private var paymentAlert: UIAlertController!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        checkoutCart = ShoppingCart.sharedInstance
        
        paymentAlert = UIAlertController(
            title: "Payment Method Not Selected",
            message: "Please Choose your payment method From the Payment Method Section",
            preferredStyle: .alert
        )
        
        let yesAction = UIAlertAction(title: "OK!", style: .default, handler: nil)
        paymentAlert.addAction(yesAction)
        
        bkashChecked = false
        rocketChecked = false
        cashOnDeliveryChecked = false
        
        defaults = UserDefaults.standard
        
        if !defaults.bool(forKey: "SeesionUserLoggedIN") {
            checkOutTotalAmount.text = ""
            firstName.text = ""
            lastName.text = ""
            email.text = ""
            phone.text = ""
            country.text = ""
            city.text = ""
            State.text = ""
            postalCode.text = ""
            address.text = ""
            bkashOutlet.isHidden = true
            rocketOutlet.isHidden = true
            cashOnDeliveryOutlet.isHidden = true
        } else {
            bkashOutlet.isHidden = false
            rocketOutlet.isHidden = false
            cashOnDeliveryOutlet.isHidden = false
            userData = defaults.dictionary(forKey: "LoggedInUsersDetail")
            
            checkOutTotalAmount.text = "$\(defaults.value(forKey: "cartTotalAmount") ?? "")"
            firstName.text = userData?["firstName"] as? String
            lastName.text = userData?["lastName"] as? String
            email.text = userData?["usersEmail"] as? String
            phone.text = userData?["phone"] as? String
            country.text = userData?["country"] as? String
            city.text = userData?["city"] as? String
            State.text = userData?["state"] as? String
            postalCode.text = userData?["postalCode"] as? String
            address.text = userData?["address"] as? String
        }
    }
    
    @IBAction func bkashAction(_ sender: Any) {
        if !bkashChecked {
            bkashOutlet.setImage(UIImage(named: "checkboxChecked-icon-40.png"), for: .normal)
            rocketOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            cashOnDeliveryOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            bkashChecked = true
            rocketChecked = false
            cashOnDeliveryChecked = false
            defaults.set("Bkash", forKey: "paymentMethodUsed")
            defaults.synchronize()
        } else {
            bkashOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            rocketOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            cashOnDeliveryOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            bkashChecked = false
            rocketChecked = false
            cashOnDeliveryChecked = false
            defaults.set("", forKey: "paymentMethodUsed")
            defaults.synchronize()
        }
    }
    
    @IBAction func rocketAction(_ sender: Any) {
        if !rocketChecked {
            rocketOutlet.setImage(UIImage(named: "checkboxChecked-icon-40.png"), for: .normal)
            bkashOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            cashOnDeliveryOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            rocketChecked = true
            bkashChecked = false
            cashOnDeliveryChecked = false
            defaults.set("Rocket", forKey: "paymentMethodUsed")
            defaults.synchronize()
        } else {
            bkashOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            rocketOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            cashOnDeliveryOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            bkashChecked = false
            rocketChecked = false
            cashOnDeliveryChecked = false
            defaults.set("", forKey: "paymentMethodUsed")
            defaults.synchronize()
        }
    }
    
    @IBAction func cashOnDeliveryAction(_ sender: Any) {
        if !cashOnDeliveryChecked {
            cashOnDeliveryOutlet.setImage(UIImage(named: "checkboxChecked-icon-40.png"), for: .normal)
            rocketOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            bkashOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            cashOnDeliveryChecked = true
            rocketChecked = false
            bkashChecked = false
            defaults.set("Cash On Delivery", forKey: "paymentMethodUsed")
            defaults.synchronize()
        } else {
            bkashOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            rocketOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            cashOnDeliveryOutlet.setImage(UIImage(named: "checkboxUnchecked-icon-40.png"), for: .normal)
            bkashChecked = false
            rocketChecked = false
            cashOnDeliveryChecked = false
            defaults.set("", forKey: "paymentMethodUsed")
            defaults.synchronize()
        }
    }
    
    func getHTMLString() -> String {
        let pdfLogoUrl = "https://apiforios.appendtech.com/logo.png"
        
        guard let filePath = Bundle.main.path(forResource: "invoice", ofType: "html"),
              let singleItemFilePath = Bundle.main.path(forResource: "single_item", ofType: "html"),
              var strHTML = try? String(contentsOfFile: filePath, encoding: .utf8) else {
            return ""
        }
        
        let itemsInCartArray = checkoutCart.itemsInCart()
        
        var allItemsHTMLArray: [String] = []
        for item in itemsInCartArray {
            if var srtItemHTML = try? String(contentsOfFile: singleItemFilePath, encoding: .utf8) {
                srtItemHTML = srtItemHTML.replacingOccurrences(of: "#productName", with: item.itemName)
                srtItemHTML = srtItemHTML.replacingOccurrences(of: "#quantity", with: NSNumber(value: item.cartAddedQuantity).stringValue)
                srtItemHTML = srtItemHTML.replacingOccurrences(of: "#price", with: NSNumber(value: item.price).stringValue)
                allItemsHTMLArray.append(srtItemHTML)
            }
        }
        let allItemsHTMLString = allItemsHTMLArray.joined(separator: "\n")
        
        strHTML = strHTML.replacingOccurrences(of: "#ITEMS#", with: allItemsHTMLString)
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "MMM d, yyyy"
        let dateString = dateFormatter.string(from: Date())
        
        let ownerInfo = "ICTcom<br>An E-commerce App Project<br>Developed BY: Rony Banik (Arko)<br>"
        strHTML = strHTML.replacingOccurrences(of: "#ownerInfo", with: ownerInfo)
        strHTML = strHTML.replacingOccurrences(of: "#appIcon", with: pdfLogoUrl)
        strHTML = strHTML.replacingOccurrences(of: "#invoiceDate", with: dateString)
        strHTML = strHTML.replacingOccurrences(of: "#cartTotal", with: defaults.value(forKey: "cartTotalAmount") as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#firstName", with: userData?["firstName"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#lastName", with: userData?["lastName"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#email", with: userData?["usersEmail"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#phone", with: userData?["phone"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#country", with: userData?["country"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#city", with: userData?["city"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#state", with: userData?["state"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#postalCode", with: userData?["postalCode"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#address", with: userData?["address"] as? String ?? "")
        strHTML = strHTML.replacingOccurrences(of: "#paymentMethod", with: defaults.value(forKey: "paymentMethodUsed") as? String ?? "")
        
        return strHTML
    }
    
    func savePDF() {
        let pageFrame = CGRect(x: 0, y: 0, width: 595.2, height: 841.8)
        
        let printPageRenderer = UIPrintPageRenderer()
        printPageRenderer.setValue(NSValue(cgRect: pageFrame), forKey: "paperRect")
        printPageRenderer.setValue(NSValue(cgRect: pageFrame), forKey: "printableRect")
        let printFormatter = UIMarkupTextPrintFormatter(markupText: getHTMLString())
        printPageRenderer.addPrintFormatter(printFormatter, startingAtPageAt: 0)
        
        let pdfData = NSMutableData()
        
        UIGraphicsBeginPDFContextToData(pdfData, .zero, nil)
        UIGraphicsBeginPDFPage()
        printPageRenderer.drawPage(at: 0, in: UIGraphicsGetPDFContextBounds())
        UIGraphicsEndPDFContext()
        
        PDFFilePath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        PDFFilePath = PDFFilePath + "/invoiceFromICTcom.pdf"
        pdfData.write(toFile: PDFFilePath, atomically: true)
        print("Filepath: \(PDFFilePath)")
        defaults.set(PDFFilePath, forKey: "pdfFilePath")
        defaults.synchronize()
    }
    
    @IBAction func finishCheckOutButton(_ sender: Any) {
        if (defaults.value(forKey: "paymentMethodUsed") as? String ?? "") == "" {
            present(paymentAlert, animated: true, completion: nil)
        } else {
            savePDF()
            sendMailWithPDF()
            defaults.set("", forKey: "cartTotalAmount")
            defaults.set("", forKey: "paymentMethodUsed")
            defaults.synchronize()
            checkoutCart.clearCart()
            performSegue(withIdentifier: "checkoutToThankyouSegue", sender: nil)
        }
    }
    
    func sendMailWithPDF() {
        if MFMailComposeViewController.canSendMail() {
            let saleMail = MFMailComposeViewController()
            saleMail.mailComposeDelegate = self
            saleMail.setSubject("Invoice From ICTcom App")
            saleMail.setMessageBody("Thanks so much for your purchase.<br>Please find the attached invoice.<br>Thanks From ICTcom", isHTML: true)
            
            if let contentsOfFile = try? Data(contentsOf: URL(fileURLWithPath: PDFFilePath)) {
                saleMail.addAttachmentData(contentsOfFile, mimeType: "application/pdf", fileName: "Invoice")
            }
            
            if let userEmail = userData?["usersEmail"] as? String {
                saleMail.setToRecipients([userEmail])
            }
            
            present(saleMail, animated: true, completion: nil)
        } else {
            print("Device can't send email! maybe Simulator")
        }
    }
    
    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        controller.dismiss(animated: true, completion: nil)
    }
}
